### PR TITLE
add retry mechanism

### PR DESCRIPTION
**Pull request by lichunzhu**
### What problem does this PR solve? 
https://github.com/pingcap/dumpling/issues/178
https://github.com/pingcap/dumpling/issues/180

### What is changed and how it works?
1. Retry dumping on failed chunks.
2. Increase `maxIdleConnsPerHost` when `config.Threads` is set. Reference: https://github.com/RedTeamPentesting/monsoon/pull/3/files#diff-3ad9aea5b78bcb9b0ca9164d4cd81a2c462dbc82d2a3e7b5477f66134c65e05cR52

### Check List 

Tests 

 - Unit test
 - Integration test

Side effects

 - Possible performance regression

Related changes

 - Need to cherry-pick to the release branch
 
### Release note

- Retry dumping on failed chunks.

### CODE REVIEW

**Comment by lichunzhu** [comment]
DNM now. Manual test shows some bugs

**Comment by lichunzhu** [comment]
I have tested using `lldb` to kill several socket connections while dumping. Dumpling works fine even in this situation.

File: v4/export/config.go [hunk 1]
```diff
@@ -133,7 +135,31 @@ func (config *Config) createExternalStorage(ctx context.Context) (storage.Extern
...
+			KeepAlive: 30 * time.Second,
+			DualStack: true,
+		}).DialContext,
+		ForceAttemptHTTP2:     true,
+		MaxIdleConns:          100,
+		MaxIdleConnsPerHost:   maxIdleConnsPerHost,
+		IdleConnTimeout:       90 * time.Second,
+		TLSHandshakeTimeout:   10 * time.Second,
+		ExpectContinueTimeout: 1 * time.Second,
+	}
```
Comment (kennytm) [inline v4/export/config.go:158]:
```suggestion
	transport := http.DefaultTransport.(*http.Transport).Clone()
	transport.MaxIdleConnsPerHost = maxIdleConnsPerHost
	httpClient.Transport = transport
```

File: v4/export/dump_test.go [hunk 2]
```diff
@@ -195,3 +198,36 @@ func (s *testDumpSuite) TestDumpTableWhereClause(c *C) {
...
+	mock.ExpectQuery("SHOW CREATE DATABASE `test`").WillReturnRows(rows)
+	showCreateTableResult := "CREATE TABLE t (a INT)"
+	rows = mock.NewRows([]string{"Table", "Create Table"}).AddRow("t", showCreateTableResult)
+	mock.ExpectQuery("SHOW CREATE TABLE `test`.`t`").WillReturnRows(rows)
+	rows = mock.NewRows([]string{"column_name", "extra"}).AddRow("id", "").AddRow("name", "")
+	mock.ExpectQuery("SELECT COLUMN_NAME").WithArgs(sqlmock.AnyArg(), sqlmock.AnyArg()).WillReturnRows(rows)
+	rows = mock.NewRows([]string{"a"}).AddRow(1)
+	mock.ExpectQuery("SELECT (.) FROM `test`.`t` LIMIT 1").WillReturnRows(rows)
+	rows = mock.NewRows([]string{"a"}).AddRow(1).AddRow(2)
+	mock.ExpectQuery("SELECT (.) FROM `test`.`t`").WillReturnError(&mysql.MySQLError{Number: errno.ErrPDServerTimeout, Message: "pd is timeout"})
```
Comment (kennytm) [inline v4/export/dump_test.go:221]: can we use `9001` in place of `errno.ErrPDServerTimeout` here? looks like a total waste to add `tidb` to go.mod just to use a single constant in a unit test.

**Review by kennytm — COMMENTED** [review]
rest LGTM

File: v4/export/dump.go [hunk 3]
```diff
@@ -250,22 +255,34 @@ func dumpDatabases(ctx context.Context, conf *Config, connectPool *connectionsPo
...
 			for _, tableIR := range tableDataIRArray {
 				tableIR := tableIR
-				g.Go(func() error {
-					conn := connectPool.getConn()
-					defer connectPool.releaseConn(conn)
-					err := tableIR.Start(ctx, conn)
+				g.Add(1)
+				go func() {
+					defer g.Done()
+					retryTime := 1
```
Comment (kennytm) [inline v4/export/dump.go:261]: wait, why can't we keep using errgroup here?

Comment (lichunzhu) [reply v4/export/dump.go]: You are right. I have rewritten this part's logic using `errgroup` again.

**Comment by mightyguava** [comment]
Yes, I built dumpling from db9e82708dfa5fa0f4bbb40eab65fab4ed708be1

**Review by kennytm — APPROVED** [review]

**Pull request merged**